sound/soc/sof/ipc4-topology.c

```diff
@@ -1434,7 +1441,18 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 				alh_copier = (struct sof_ipc4_copier *)dai->private;
 				alh_data = &alh_copier->data;
 				blob->alh_cfg.mapping[i].alh_id = alh_data->gtw_cfg.node_id;
-				blob->alh_cfg.mapping[i].channel_mask = ch_mask;
+				/*
+				 * Set the same channel mask for playback as the audio data is
+				 * duplicated for all speakers. For capture, split the channels
+				 * among the aggregated DAIs. For example, with 4 channels on 2
+				 * aggregated DAIs, the channel_mask should be 0x3 and 0xc for the
+				 * two DAI's.
```

Comment:
ok, but how is the order between amplifiers defined? Is this in the dailink definition or the topology, or both?

Completely lost.

Reply:
@plbossart does the order matter at all though? From what I can tell from this code, if there are 4 channels in the feedback, 2 come from each DAI.

Reply:
yes it does, because the channels are provided to userspace or a process on the DSP, which HAS to know which channel is what.

Reply:
@plbossart The lower bits will be assigned to the 2nd dai. For example, the dai that associate to rt1316-2's channel_mask will be 0x3 and the dai that associate to rt1316-1's channel_mask will be 0xc.
The magic is that `sof_widget_ready` will call `sof_connect_dai_widget` to connect the cpu_dai to the widget and add the widget to `sdev->widget_list`.
Then we assign channel_mask in the `list_for_each_entry(w, &sdev->widget_list, list)` loop. Therefore, the 2nd dai's channel_mask will be assigned first.

Reply:
@bardliao you need to document that convention with a comment in the code. The channel masks used depend on the cpu_dais used in the dailink at the machine driver level, which actually comes from the tables in soc_acpi files depending on the _ADR and devID registers for each codec.

Reply:
> @bardliao you need to document that convention with a comment in the code. The channel masks used depend on the cpu_dais used in the dailink at the machine driver level, which actually comes from the tables in soc_acpi files depending on the _ADR and devID registers for each codec.

@plbossart Added the comment above in the code.